C/common/reading.cpp

```diff
@@ -283,3 +267,70 @@ ostringstream assetTime;
 	}
 
 }
+
+/**
+ * Set the system timestamp from a string of the format
+ * 2019-01-01 10:00:00.123456+08:00
+ * The timeval is populate in UTC
+ *
+ * @param timestamp	The timestamp string
+ */
+void Reading::setTimestamp(const string& timestamp)
+{
+	stringToTimestamp(timestamp, &m_timestamp);
+}
+
+/**
+ * Set the user timestamp from a string of the format
+ * 2019-01-01 10:00:00.123456+08:00
+ * The timeval is populate in UTC
+ *
+ * @param timestamp	The timestamp string
+ */
+void Reading::setUserTimestamp(const string& timestamp)
+{
+	stringToTimestamp(timestamp, &m_userTimestamp);
+}
+
+/**
+ * Convert a string timestamp, with milliseconds to a 
+ * struct timeval.
+ *
+ * Timezone handling
+ *    The timezone in the string is extracted to get UTC values
+ *    This is then converted to the local timezone as readings 
+ *    are always expressed in the localtime zone
+ *
+ * @param timestamp	String timestamp
+ * @param ts		Struct timeval to populate
+ */
+void Reading::stringToTimestamp(const string& timestamp, struct timeval *ts)
+{
+	struct tm tm;
+	strptime(timestamp.c_str(), "%Y-%m-%d %H:%M:%S", &tm);
+	ts->tv_sec = mktime(&tm);
+	const char *ptr = timestamp.c_str();
+	while (*ptr && *ptr != '.')
+		ptr++;
+	if (*ptr)
+	{
+		ptr++;
+		ts->tv_usec = strtol(ptr, NULL, 10);
+	}
+	// Get the timezone from the string and convert to UTC
+	ptr = timestamp.c_str() + 10; // Skip date as it contains '-' characters
+	while (*ptr && *ptr != '-' && *ptr != '+')
+                ptr++;
+	if (*ptr)
+	{
+		int h, m;
+		int sign = (*ptr == '+' ? -1 : +1);
+		ptr++;
+		sscanf(ptr, "%02d:%02d", &h, &m);
+		ts->tv_sec += sign * ((3600 * h) + (60 * m));
+	}
+	// Now convert to the local time
+	extern long timezone;
+	tzset();
+	ts->tv_sec -= timezone;
```

Comment:
Will this be subtraction or addition? I think local time is "timezone" seconds ahead of UTC.

Reply:
Yes, for timezones east of Greenwich timezone is a negative number and timezones west of Greenwich it is a positive number. 

```
foglamp@foglamp-18:~$ cat t.c
main()
{
	extern long timezone;
	tzset();
	printf("Timezone: %ld\n", timezone);
}
foglamp@foglamp-18:~$ export TZ=EET
foglamp@foglamp-18:~$ ./t
Timezone: -7200
foglamp@foglamp-18:~$ export TZ=CET
foglamp@foglamp-18:~$ ./t
Timezone: -3600
foglamp@foglamp-18:~$ export TZ=EST
foglamp@foglamp-18:~$ ./t
Timezone: 18000
foglamp@foglamp-18:~$ 
```

Reply:
Given the nature of this issue it would be good if you could try it in all your time zones.

Reply:
Ok, got it. Thanks.